api/src/services/schema.ts

```diff
@@ -0,0 +1,290 @@
+					);
+					if (existingField) {
+						throw new InvalidPayloadException(
+							`Provided diff is trying to create field "${field}" but it already exists. Please regenerate a new diff and try again.`
+						);
+					}
+				} else if (diffField.diff[0]?.kind === 'D') {
+					const existingField = snapshotWithHash.fields.find(
+						(f) => f.collection === diffField.collection && f.field === diffField.field
+					);
+					if (!existingField) {
+						throw new InvalidPayloadException(
+							`Provided diff is trying to delete field "${field}" but it does not exist. Please regenerate a new diff and try again.`
+						);
+					}
+				} else {
+					const matchingField = snapshotWithHash.fields.find((f) => f.hash === diffField.hash);
+					if (!matchingField) {
+						throw new InvalidPayloadException(
+							`Provided diff to update field "${field}" does not match the current instance's field, indicating it has changed after this diff was generated. Please regenerate a new diff and try again.`
+						);
+					}
+				}
+			}
+			for (const diffRelation of payload.diff.relations) {
+				let relation = `${diffRelation.collection}.${diffRelation.field}`;
+				if (diffRelation.related_collection) relation += `-> ${diffRelation.related_collection}`;
+
+				if (diffRelation.diff[0]?.kind === 'N') {
+					const existingRelation = snapshotWithHash.relations.find(
+						(r) => r.collection === diffRelation.collection && r.field === diffRelation.field
+					);
+					if (existingRelation) {
+						throw new InvalidPayloadException(
+							`Provided diff is trying to create relation "${relation}" but it already exists. Please regenerate a new diff and try again.`
+						);
+					}
+				} else if (diffRelation.diff[0]?.kind === 'D') {
+					const existingRelation = snapshotWithHash.relations.find(
+						(r) => r.collection === diffRelation.collection && r.field === diffRelation.field
+					);
+					if (!existingRelation) {
+						throw new InvalidPayloadException(
+							`Provided diff is trying to delete relation "${relation}" but it does not exist. Please regenerate a new diff and try again.`
+						);
+					}
+				} else {
+					const matchingRelation = snapshotWithHash.relations.find((r) => r.hash === diffRelation.hash);
+					if (!matchingRelation) {
+						throw new InvalidPayloadException(
+							`Provided diff for relation "${relation}" does not match the current instance's relation, indicating it has changed after this diff was generated. Please regenerate a new diff and try again.`
+						);
+					}
+				}
+			}
+
+			throw new InvalidPayloadException(
+				`Provided hash "${payload.hash}" does not match the current instance's schema hash "${snapshotWithHash.hash}". Please regenerate a new diff and try again.`
+			);
+		}
+
+		if (
+			payload.diff.collections.length === 0 &&
+			payload.diff.fields.length === 0 &&
+			payload.diff.relations.length === 0
+		) {
+			return;
+		}
```

Comment:
Should this check be first, so we can skip the validation step if there's no diff to be applied?